File: servermock/jqpmserver/github.py
"""
GitHub API calls.
"""
import json
from urllib import quote, urlencode
from base64 import b64decode
from StringIO import StringIO

import requests


GITHUB_URL = "https://api.github.com"


def search_repos(query):
    """Search GitHub for repositories matching given query.
    Returns a list of dicts.
    """
    url = "%s/legacy/repos/search/%s" % (GITHUB_URL, quote(query))
    response = requests.get(url)

    repos = json.loads(response.content)
    return repos['repositories']


def list_files(user, repo, ref='master'):
    """Lists all files in root directory of ``repo`` belonging
    to given ``user``.
    Returns a dictionary mapping file and directory names to URLs.
    """
    url = "%s/repos/%s/%s/git/trees/%s" % (GITHUB_URL, user, repo, ref)
    response = requests.get(url)

    tree_info = json.loads(response.content)
    return dict(
        ("%s%s" % (t['path'], "/" if t['type'] == 'tree' else ''), t['url'])
        for t in tree_info['tree']
    )


def get_commits(user, repo, from_ref=None, path=None):
    """Gets list of commits from ``repo`` belonging to given ``user``,
    optionally filtering by ``path`` and/or starting from given ``ref``.
    Returns a list of dictionaries.
    """
    args = {}
    if from_ref is not None:
        args['sha'] = from_ref
    if path is not None:
        args['path'] = path

    url = "%s/repos/%s/%s/commits" % (GITHUB_URL, user, repo)
    if args:
        url += "?%s" % urlencode(args)
    response = requests.get(url)

    return json.loads(response.content)


def retrieve_file(user, repo, path, ref='master'):
    """Retrieves contents of file under ``path`` from a ``repo``
    that belongs to given ``user``.
    Returns a file-like object with file content.
    """
    if path.startswith('/'):
        path = path[1:]

    url = "%s/repos/%s/%s/contents/%s" % (GITHUB_URL, user, repo, path)
    url += "?%s" % urlencode({'ref': ref})
    response = requests.get(url)

    file_info = json.loads(response.content)
    return StringIO(b64decode(file_info['content']))

File: lib/github.js
/**
 * Module for GitHub API calls.
 */

'use strict';


var util = require('util')
  , buffer = require('buffer')
  , querystring = require('querystring')
  , http = require('http')
  , https = require('https')
  , url = require('url')
  ;


var GITHUB_URL = "https://api.github.com";


module.exports = {

	/**
	 * Search for repositories matching given query.
	 * :return: An array of objects with repository infos
	 */
	searchRepos: function(query, callback) {
		var url = util.format("%s/legacy/repos/search/%s",
							  GITHUB_URL, querystring.escape(query));
		httpGet(url, function(err, respBody) {
			if (err) {
				callback(err);
			} else {
				var resp = JSON.parse(respBody);
				callback(null, resp.repositories);
			}
		});
	},

	/**
	 * Lists all files in root directory of ``repo`` belonging
	 * to given ``user``.
	 * :return: Object mapping file and directory names to URLs
	 */
	listFiles: function(user, repo, ref, callback) {
		if (!callback) {
			callback = ref;
			ref = 'master';	// ``ref`` is optional
		}

		var url = util.format("%s/repos/%s/%s/git/trees/%s",
							  GITHUB_URL, user, repo, ref);
		httpGet(url, function(err, respBody) {
			if (err) {
				callback(err);
			} else {
				var resp = JSON.parse(respBody);
				var res = {};
				resp.tree.forEach(function(item) {
					var key = item.path;
					if (item.type === 'tree')
						key += '/';
					res[key] = item.url;
				});
				callback(null, res);
			}
		});
	},

	/**
	 * Gets list of commits from ``repo`` belonging to given ``user``.
	 * :param options: May contain:
	 *                 * ``fromRef`` to specify starting revision
	 *                 * ``path`` includes only commits touching the path
	 * :return: Array of objects with commit infos
	 */
	getCommits: function(user, repo, options, callback) {
		if (!callback) {
			callback = options;
			options = {}; // ``options`` are optional
		}

		var args = {
			sha: (options.fromRef ? options.fromRef : 'master')
		};
		if (options.path)
			args.path = options.path;
		var qsArgs = querystring.encode(args);

		var url = util.format("%s/repos/%s/%s/commits",
							  GITHUB_URL, user, repo);
		if (qsArgs)
			url += "?" + qsArgs;

		httpGet(url, function(err, respBody) {
			if (err) {
				callback(err);
			} else {
				var commits = JSON.parse(respBody);
				callback(null, commits);
			}
		});
	},

	/**
	 * Download file from given URL that points to GitHub file.
	 */
	downloadFile: function(url, callback) {
		httpGet(url, function(err, respBody) {
			if (err) {
				callback(err);
			} else {
				var resp = JSON.parse(respBody);
				var file = base64Decode(resp.content);
				callback(null, file);
			}
		});
	},
};


function httpGet(targetUrl, callback) {
	if (typeof targetUrl === 'string')
		targetUrl = url.parse(targetUrl);
	var protocol = targetUrl.protocol === 'https:' ? https : http;
	var req = protocol.get(targetUrl, function(resp) {
		if (resp.statusCode >= 200 && resp.statusCode < 300) {
			resp.setEncoding('utf8');

			var respBody = "";
			resp.on('data', function(chunk) {
				respBody += chunk;
			});
			resp.on('end', function() {
				callback(null, respBody);
			});
		} else {
			var err = new Error(util.format("HTTP error: %s", resp.statusCode));
			err.statusCode = resp.statusCode;
			callback(err);
		}
	});
	req.on('error', callback);
}

function base64Decode(data) {
	return buffer.Buffer(data, 'base64').toString();
}
